File: leetcode_by_cpp/46.全排列.cpp
/*
 * @lc app=leetcode.cn id=46 lang=cpp
 *
 * [46] 全排列
 *
 * https://leetcode-cn.com/problems/permutations/description/
 *
 * algorithms
 * Medium (76.70%)
 * Likes:    878
 * Dislikes: 0
 * Total Accepted:    187.8K
 * Total Submissions: 244.5K
 * Testcase Example:  '[1,2,3]'
 *
 * 给定一个 没有重复 数字的序列，返回其所有可能的全排列。
 * 
 * 示例:
 * 
 * 输入: [1,2,3]
 * 输出:
 * [
 * ⁠ [1,2,3],
 * ⁠ [1,3,2],
 * ⁠ [2,1,3],
 * ⁠ [2,3,1],
 * ⁠ [3,1,2],
 * ⁠ [3,2,1]
 * ]
 * 
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
private:
    vector<vector<int>> res;

public:
    void backtracking(vector<int> &nums, int begin)
    {
        if (begin == nums.size())
            res.push_back(nums);
        else
        {
            for (int i = begin; i < nums.size(); i++)
            {
                swap(nums[begin], nums[i]); //交换法
                backtracking(nums, begin + 1);
                swap(nums[begin], nums[i]);
            }
        }
    }
    vector<vector<int>> permute(vector<int> &nums)
    {
        backtracking(nums, 0);
        return res;
    }
};
// @lc code=end

File: leetcode_by_rust/src/bin/46.全排列.rs
/*
 * @lc app=leetcode.cn id=46 lang=rust
 *
 * [46] 全排列
 *
 * https://leetcode.cn/problems/permutations/description/
 *
 * algorithms
 * Medium (78.86%)
 * Likes:    2565
 * Dislikes: 0
 * Total Accepted:    868.6K
 * Total Submissions: 1.1M
 * Testcase Example:  '[1,2,3]'
 *
 * 给定一个不含重复数字的数组 nums ，返回其 所有可能的全排列 。你可以 按任意顺序 返回答案。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：nums = [1,2,3]
 * 输出：[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
 *
 *
 * 示例 2：
 *
 *
 * 输入：nums = [0,1]
 * 输出：[[0,1],[1,0]]
 *
 *
 * 示例 3：
 *
 *
 * 输入：nums = [1]
 * 输出：[[1]]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 1 <= nums.length <= 6
 * -10 <= nums[i] <= 10
 * nums 中的所有整数 互不相同
 *
 *
 */

// @lc code=start
pub fn back_tracking(nums: &mut Vec<i32>, begin: usize, res: &mut Vec<Vec<i32>>) {
    if begin == nums.len() {
        res.push(nums.to_vec());
    }
    for i in begin..nums.len() {
        nums.swap(i, begin);
        Self::back_tracking(nums, begin + 1, res);
        nums.swap(i, begin);
    }
}
impl Solution {
    pub fn permute(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut res: Vec<Vec<i32>> = Vec::new();
        back_tracking(&mut nums, 0, &mut res);
        res
    }
}
// @lc code=end
